src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+    this.map.resize();
+};
+
+proto.resolveOnRender = function(resolve) {
+    var map = this.map;
+
+    map.on('render', function onRender() {
+        if(map.loaded()) {
+            map.off('render', onRender);
+            resolve();
+        }
+    });
+};
+
+proto.rejectOnError = function(reject) {
+    var map = this.map;
+
+    function handler() {
+        reject(new Error(constants.mapOnErrorMsg));
+    }
+
+    map.once('error', handler);
+    map.once('style.error', handler);
+    map.once('source.error', handler);
+    map.once('tile.error', handler);
+    map.once('layer.error', handler);
+};
+
+proto.createFramework = function(fullLayout) {
+    var self = this;
+
+    var div = self.div = document.createElement('div');
+
+    div.id = self.uid;
+    div.style.position = 'absolute';
+
+    self.container.appendChild(div);
+
+    // create mock x/y axes for hover routine
+
+    self.xaxis = {
+        _id: 'x',
+        c2p: function(v) { return self.project(v).x; }
+    };
+
+    self.yaxis = {
+        _id: 'y',
+        c2p: function(v) { return self.project(v).y; }
+    };
+
+    self.updateFramework(fullLayout);
+};
+
+proto.updateFramework = function(fullLayout) {
+    var domain = fullLayout[this.id].domain,
+        size = fullLayout._size;
+
+    var style = this.div.style;
+
+    // TODO Is this correct? It seems to get the map zoom level wrong?
+
+    style.width = size.w * (domain.x[1] - domain.x[0]) + 'px';
+    style.height = size.h * (domain.y[1] - domain.y[0]) + 'px';
+    style.left = size.l + domain.x[0] * size.w + 'px';
+    style.top = size.t + (1 - domain.y[1]) * size.h + 'px';
+
+    this.xaxis._offset = size.l + domain.x[0] * size.w;
+    this.xaxis._length = size.w * (domain.x[1] - domain.x[0]);
+
+    this.yaxis._offset = size.t + (1 - domain.y[1]) * size.h;
+    this.yaxis._length = size.h * (domain.y[1] - domain.y[0]);
+};
+
+proto.updateLayers = function() {
+    var opts = this.opts,
+        layers = opts.layers,
+        layerList = this.layerList,
+        i;
+
+    // if the layer arrays don't match,
+    // don't try to be smart,
+    // delete them all, and start all over.
+
+    if(layers.length !== layerList.length) {
+        for(i = layerList.length - 1; i > -1; i--) {
+            layerList[i].dispose();
+            layerList.pop();
+        }
+
+        for(i = 0; i < layers.length; i++) {
+            layerList.push(createMapboxLayer(this, i, layers[i]));
+        }
+    }
+    else {
+        for(i = 0; i < layers.length; i++) {
+            layerList[i].update(layers[i]);
+        }
+    }
+};
+
+proto.destroy = function() {
+    this.map.remove();
+    this.container.removeChild(this.div);
+};
+
+proto.toImage = function() {
+    return this.map.getCanvas().toDataURL();
+};
+
+// convenience wrapper to create blank GeoJSON sources
+// and avoid 'invalid GeoJSON' errors
+proto.createGeoJSONSource = function() {
+    var blank = {
+        type: 'Feature',
```

Comment:
No. But changing the GeoJSON type using `map.setData` (as [here](https://github.com/plotly/plotly.js/pull/626/commits/39684178450ca1a25671bd2d24b0f0ad82e8eb94#diff-95de105f2fc614fa15caf457477c452eR85) is easy enough.